File: camount-swift/Sources/Camount/Core/SanitizeInput.swift
import Foundation

struct SanitizedInput {
    let text: String
    let cursor: Int
}

func sanitizeInput(text: String, cursor: Int, config: AmountConfig) -> SanitizedInput {
    let src = Array(text)
    var builder = ""
    var separatorSeen = false
    var integerDigits = 0
    var fractionDigits = 0
    let originalCursor = max(0, min(cursor, src.count))
    var mappedCursor = 0

    for i in 0..<src.count {
        let c = src[i]
        var kept = false
        if config.isDigit(c) {
            let underLimit =
                separatorSeen
                ? fractionDigits < config.maximumFractionDigits
                : integerDigits < config.maximumNotationDigits
            if underLimit {
                builder.append(c)
                if separatorSeen { fractionDigits += 1 } else { integerDigits += 1 }
                kept = true
            }
        } else if config.isInputSeparator(c) && !separatorSeen && config.maximumFractionDigits > 0 {
            separatorSeen = true
            builder.append(config.decimalSeparator)
            kept = true
        }
        if kept && i < originalCursor { mappedCursor += 1 }
    }

    return SanitizedInput(text: builder, cursor: min(mappedCursor, builder.count))
}

File: camount-js/src/core/sanitizeInput.ts
import type { AmountConfig, SanitizedInput } from "./types";
import { configIsDigit, configIsInputSeparator } from "./types";

export function sanitizeInput(
  text: string,
  cursor: number,
  config: AmountConfig,
): SanitizedInput {
  const src = [...text];
  let builder = "";
  let separatorSeen = false;
  let integerDigits = 0;
  let fractionDigits = 0;
  const originalCursor = Math.max(0, Math.min(cursor, src.length));
  let mappedCursor = 0;

  for (let i = 0; i < src.length; i++) {
    const c = src[i]!;
    let kept = false;
    if (configIsDigit(config, c)) {
      const underLimit = separatorSeen
        ? fractionDigits < config.maximumFractionDigits
        : integerDigits < config.maximumNotationDigits;
      if (underLimit) {
        builder += c;
        if (separatorSeen) fractionDigits += 1;
        else integerDigits += 1;
        kept = true;
      }
    } else if (
      configIsInputSeparator(config, c) &&
      !separatorSeen &&
      config.maximumFractionDigits > 0
    ) {
      separatorSeen = true;
      builder += config.decimalSeparator;
      kept = true;
    }
    if (kept && i < originalCursor) mappedCursor += 1;
  }

  return { text: builder, cursor: Math.min(mappedCursor, builder.length) };
}
